tests: Add libraries and multicast uplink automation

kuldeepkash opened the pull request:
1. Adding library support to create and verify multicast PIM multi-uplink
    topology and configuration
2. Adding new testcases to test multicast uplink functionality end to end.

Signed-off-by: Kuldeep Kashyap <kashyapk@vmware.com>

donaldsharp commented:
Why are we submitting removal of functions from our test suites.  It's clear we are using these functions and the test failures show that it would have been caught if the topotests were run first.  Why are we not even running our code?  This does not inspire much confidence here at all.

donaldsharp reviewed (changes requested):
NACK.  Please do not just arbitrarily remove functionality from our test suites

kuldeepkash commented:
> Why are we submitting removal of functions from our test suites. It's clear we are using these functions and the test failures show that it would have been caught if the topotests were run first. Why are we not even running our code? This does not inspire much confidence here at all.

@donaldsharp , Thanks for looking into it, there is a new API written already (library: */lib/pim.py API name: scapy_send_bsr_raw_packet), now we are using this newly written API. 

Note: I have updated PR and added back old API, lets keep it in codebase.

kuldeepkash commented:
> NACK. Please do not just arbitrarily remove functionality from our test suites

As there was a new API written already so it got removed. But for now I have updated PR and added back old API, lets keep it in codebase.

This PR is to add new APIs needed for multicast uplink automation. I will be adding tests to the same PR.

kuldeepkash commented:
CI: rerun

donaldsharp commented:
This is a NACK.  Submitting a PR that has topotest changes to allow you to configure cli in FRR that is not available in upstream is just a No from me.  Why is our time being wasted with this?

donaldsharp commented on tests/topotests/lib/pim.py:
```diff
@@ -405,76 +405,6 @@ def find_rp_details(tgen, topo):
     return rp_details
 
 
-def configure_pim_force_expire(tgen, topo, input_dict, build=False):
```
and now in the *third* version of this commit we remove a function that has absolutely nothing to do with the changes in the commit message.  This removal wasn't in the first two versions of this commit.  Stop playing games, I'm done with this whole thing.  You need to figure out how to properly break up your commits as well as properly test your code before you ask the community to review.  You are wasting our time.

kuldeepkash replied:
@donaldsharp , We discussed, I will update the PR on the basis of whatever clis are not used or needed in community specially for multicast. I did this change because we are updating pim.py file in this PR.

I did a thorough check and this API configure_pim_force_expire is no where used. I thought you will ok to remove all the APIs, which are using not needed CLIs in community. Please understand this is not to waste anyone's time. Just trying to enhance community's test capabilities. 

If you are not ok with the changes and tests added, we can close this PR and raise another PR post vacation. will keep note of your comments going forward. Thank you.

donaldsharp reviewed (changes requested):
NACK

kuldeepkash commented on tests/topotests/lib/common_config.py:
```diff
@@ -4985,3 +4985,52 @@ def scapy_send_raw_packet(tgen, topo, senderRouter, intf, packet=None):
     return True
+
+
+def kernel_requires_l3mdev_adjustment():
```
@donaldsharp , This function is not being added in this PR, it is already present, written by someone else from community.  It is showing up only because of some white space addition.
